### PR TITLE
Create Cancel Payment command (again)

### PR DESCRIPTION
**Pull request by Gonals**
This unreverts https://github.com/Expensify/App/pull/32674 with a couple of changes to fix the issues found during QA

### Fixed Issues
$ https://github.com/Expensify/App/issues/32426
$ https://github.com/Expensify/App/issues/32427
$ https://github.com/Expensify/App/issues/33517
$ https://github.com/Expensify/App/issues/33509
$ https://github.com/Expensify/App/issues/33507
PROPOSAL:

### Tests
**Pay elsewhere**
1. From an account (account A) create a new workspace and invite account B to it.
2. Log into account B and request money from the new workspace.
3. Log back into account A and Pay the money request. Select "Pay elsewhere" It should end up like this:
Screenshot 2023-12-13 at 5 22 55 PM

4. Confirm you now see the 3dot menu on the top right. Click on it:
Screenshot 2023-12-13 at 5 23 42 PM

5. Select the cancel payment option. A modl should appear. Confirm the cancellation:
Screenshot 2023-12-13 at 5 23 49 PM

7. The payment should now be cancelled and you should see a message indicating so. You should now be able to pay again (and you can repeat this process a bunch of times):
Screenshot 2023-12-13 at 5 25 16 PM

**Pay with expensify**
- Make sure your admin has a valid withdrawal bank account, and your user a valid deposit one
- Repeat the previous flow, but using "Pay with Expensify" instead of "Pay elsewhere"
- Confirm it basically behaves the same way

- [x] Verify that no errors appear in the JS console

### Offline tests
Same as tests, but offline

### QA Steps
Same as tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by Gonals** [comment]
OFF HOLD!

**Comment by Gonals** [comment]
Bump @Ollyws!

**Comment by Gonals** [comment]
> `Cancel payment` in the three dots menu isn't translated into Spanish: Screenshot 2024-01-10 at 15 32 48

Fixed @Ollyws!

**Comment by Gonals** [comment]
@Ollyws, can you make sure you make this one of your top priorities? It has been stuck for a while and we want it merged soon.

Thanks!

**Review by Ollyws — APPROVED** [review]
I wasn't able to test the `Pay with expensify` flow as I don't have an account with a valid bank account, but everything else LGTM.

**Comment by rlinoz** [comment]
Code looks good!

Just let me confirm how things should behave

Going back and forth with canceling/paying updates previous messages(look at pic2), should that happen?
Screenshot 2024-01-12 at 11 05 47

Cancelling the payment changes the message to look like the admins requested money from the user, which is not what is in your screenshot, I am running latest Web-E and Auth main, maybe something changed?
Screenshot 2024-01-12 at 11 05 56

When I refresh the page the message `You marked this report as reimbursed` appears and the `Canceled` tag disappears:
Screenshot 2024-01-12 at 11 06 27

Let me know if I can help!

**Comment by Gonals** [comment]
> Going back and forth with canceling/paying updates previous messages(look at pic2), should that happen?
Yep. Maybe we can discuss some polishing down the line, but, for now, tht's the plan.

> Cancelling the payment changes the message to look like the admins requested money from the user, which is not what is in your screenshot, I am running latest Web-E and Auth main, maybe something changed?

Hmm. This does seem to be missing something. I'll take a look

> When I refresh the page the message You marked this report as reimbursed appears and the Canceled tag disappears:

This seems wrong too, but it might need a backend fix. Checking it out

**Comment by Gonals** [comment]
@rlinoz, seems like someone had messed up the money request action messages. Should be fixed now!

**Comment by Gonals** [comment]
> The messages are correct now, but still flip when you change the status, checking the onyx responses it doesn't seem like backend, but I am not 100% sure, can't find what would flip them in your PR either.
> 
>  Screen.Recording.2024-01-12.at.14.22.28.mov

Ah, that message (paid/owes) should flip. We are good 👍

**Review by rlinoz — APPROVED** [review]

File: src/libs/actions/IOU.js [hunk 1]
```diff
@@ -3328,6 +3328,109 @@ function submitReport(expenseReport) {
...
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${expenseReport.reportID}`,
+            value: {
+                ...expenseReport,
+                lastMessageText: lodashGet(optimisticReportAction, 'message.0.text', ''),
+                lastMessageHtml: lodashGet(optimisticReportAction, 'message.0.html', ''),
+                state: isFree ? CONST.REPORT.STATE.SUBMITTED : CONST.REPORT.STATE.OPEN,
+                stateNum: isFree ? CONST.REPORT.STATE_NUM.PROCESSING : CONST.REPORT.STATE.OPEN,
+                statusNum: isFree ? CONST.REPORT.STATUS.SUBMITTED : CONST.REPORT.STATE.OPEN,
```
Comment (VickyStash) [inline src/libs/actions/IOU.js:3359]: @Gonals @rlinoz I'm working on the `IOU` file [TS migration](https://github.com/Expensify/App/pull/34092), and after merging the latest main I think I've found an issue in this new part.
I guess it should be
```
stateNum: isFree ? CONST.REPORT.STATE_NUM.PROCESSING : CONST.REPORT.STATE_NUM.OPEN,
statusNum: isFree ? CONST.REPORT.STATUS.SUBMITTED : CONST.REPORT.STATUS.OPEN,
```
So using `CONST.REPORT.STATE_NUM.OPEN` instead of `CONST.REPORT.STATE.OPEN` for `stateNum`.
And using `CONST.REPORT.STATUS.OPEN` instead of `CONST.REPORT.STATE.OPEN` for `statusNum`

What do you think?

Comment (Gonals) [reply src/libs/actions/IOU.js:3359]: I think you are right, yeah. Thanks for catching that! I'll create a PR

File: src/pages/home/report/ReportActionItemMessage.tsx [hunk 2]
```diff
@@ -57,7 +57,7 @@ function ReportActionItemMessage({action, displayAsGroup, reportID, style, isHid
         const originalMessage = action.actionName === CONST.REPORT.ACTIONS.TYPE.IOU ? action.originalMessage : null;
         const iouReportID = originalMessage?.IOUReportID;
         if (iouReportID) {
-            iouMessage = ReportUtils.getReportPreviewMessage(ReportUtils.getReport(iouReportID), action);
+            iouMessage = ReportUtils.getIOUReportActionDisplayMessage(action);
```
Comment (neil-marcellini) [inline src/pages/home/report/ReportActionItemMessage.tsx:60]: @Gonals Why was this changed? I think it caused this deploy blocker https://github.com/Expensify/App/issues/34685
